### ISSUE TITLE
Hyperparameters obtained from Hyperopt extremely unstable 

### ISSUE THREAD

**Issue by Sanchita-P** [hyperopt/hyperopt#642]
Hey there, 

Hope you're staying safe and healthy! I required your help with understanding why the hyperparamters obtained from Hyperopt are extremely unstable (very different across different runs). I have worked on exploring different space ranges, even the data-sets are exactly the same in each run. However, each iteration of running the code gives me completely different optimal hyperparameters. Find below the code used by me, I would value your feedback and advice on how I can solve this as well as make my code better. Thanks a lot!

PS: I don't want to put a seed here to ensure the same results are repeated. I want to solve the inherent problem! :)

Overall summary: Train and test split -> Determining the hyperparameters using cross-validation with the train data -> Training with entire train data and performance metrics with test data

```
x_train, x_test, y_train, y_test = train_test_split(df_modelling[cols_best_rfe], labels, test_size=0.25,random_state=16)
n_folds = 5
lgb_train = lgb.Dataset(x_train, y_train)

def objective(params, n_folds = n_folds):
    """Objective function for Gradient Boosting Machine Hyperparameter Tuning"""    
    params['objective'] = 'binary'
    print(params)
    params['max_depth'] = int(params['max_depth'])
    params['num_leaves'] = int(params['num_leaves'])
    params['min_child_samples'] = int(params['min_child_samples'])
    params['subsample_freq'] = int(params['subsample_freq'])

    # Perform n_fold cross validation with hyperparameters

    # Use early stopping and evalute based on ROC AUC
    cv_results = lgb.cv(params, lgb.Dataset(x_train, y_train), nfold=n_folds, num_boost_round=10000, 
                        early_stopping_rounds=100, metrics='auc')
  
    # Extract the best score
    best_score = max(cv_results['auc-mean'])
    
    # Loss must be minimized
    loss = 1 - best_score
    num_iteration = int(np.argmax(cv_results['auc-mean']) + 1)
    
    of_connection = open(out_file, 'a')
    writer = csv.writer(of_connection)
    writer.writerow([loss, params, num_iteration])
    
    # Dictionary with information for evaluation
    return {'loss': loss, 'params': params, 'status': STATUS_OK, 'estimators': num_iteration}

space = {
    'min_child_samples': hp.quniform('min_child_samples', 5, 100, 5), 
    'reg_alpha': hp.uniform('reg_alpha', 0.0, 1.0),
    'reg_lambda': hp.uniform('reg_lambda', 0.0, 1.0),
    'colsample_bytree': hp.uniform('colsample_bytree', 0.5, 1.0),
    'max_depth' : hp.quniform('max_depth', 3, 10, 1),
    'subsample' : hp.quniform('subsample', 0.6, 1, 0.05),
    'num_leaves': hp.quniform('num_leaves', 20, 150, 1),  
    'subsample_freq': hp.quniform('subsample_freq',0,10,1),
    'min_gain_to_split': hp.quniform('min_gain_to_split', 0.01, 0.1, 0.01),                        
    'learning_rate' : hp.choice('learning_rate', [0.05])                               
}

out_file = 'results/gbm_trials_best_v1.csv'
of_connection = open(out_file, 'w')
writer = csv.writer(of_connection)

writer.writerow(['loss', 'params', 'estimators'])
of_connection.close()

trials = Trials()
best = fmin(objective, space, algo=tpe.suggest, trials=trials, max_evals=10)

results = pd.read_csv('results/gbm_trials_best_v1.csv')

# Sort with best scores on top and reset index for slicing
results.sort_values('loss', ascending = True, inplace = True)
results.reset_index(inplace = True, drop = True)
results.head()
best_bayes_estimators = int(results.loc[0, 'estimators'])

best['max_depth'] = int(best['max_depth'])
best['num_leaves'] = int(best['num_leaves'])
#best['subsample_for_bin'] = int(best['subsample_for_bin'])
best['min_child_samples'] = int(best['min_child_samples'])
#best['num_iteration'] = int(best_bayes_estimators * 1.1)
num_boost_round=int(best_bayes_estimators * 1.1)
best['objective'] = 'binary'
best['boosting_type'] = 'gbdt'
best['learning_rate'] = 0.05
best['subsample_freq'] = int(best['subsample_freq'])

lgb_train = lgb.Dataset(x_train, y_train)
best_gbm = lgb.train(params=best, train_set=lgb_train, num_boost_round=num_boost_round)
```

**Comment by bjkomer** [comment]
One reason I can see is that your `max_evals` is relatively low. By default `tpe` selects the first 20 hyperparameters randomly as a warm up to get an idea of where to start. If you only try 10 configurations they will all be chosen randomly, which explains why your results are so different each time. See #632 for an example of how to change this default. If you want more consistent results in general, you will have to make `max_evals` much larger.

**Comment by Sanchita-P** [comment]
Hi @bjkomer , you raise a valid point. I was unaware that the first 20 iterations are chosen randomly. This sounds like a strong hypothesis for the variance and randomness in the observed hyperparameters. Currently, the performance of the model can also suffer due to the random hyperparameters. 
Based on your experience, what is the **minimum number of max_evals that should be chosen?** Is there a methodology to follow for this?
I didn't understand your referenced post (632) for changing this default methodology and why would we do so?

Thanks!

**Comment by bjkomer** [comment]
I'm not aware of a good methodology for picking `max_evals`, but it will depend on the size of the search space, the complexity of the problem, and how long you are willing to wait. There is no guarantee that the 'perfectly optimal' configuration will be found, but in general later evaluations will tend to perform better than earlier ones, so you will want to pick `max_evals` to be as high as you can given your time constraint. For example, if your model takes a few seconds to train you can easily run thousands of evaluations, but if your model takes hours to train you can't go that high without some kind of parallelism. In a project I worked on in the past I used `max_evals=500` on a search space with 9 hyperparameters (similar size to yours) and got good results. I left the default random number of startup jobs at 20.

In general I would suggest running the evaluation iteratively and saving the results each step. This way you don't need to choose how long to run it for ahead of time and can stop and continue where you left off if needed. Something like this:

```python
from hyperopt import fmin, tpe
import pickle  # optional way of saving results

# ...

# Initialize an empty trials database, or reload where you left off
if reload_results == True:
    trials = pickle.load(open("my_results.pkl", "rb"))
else:
    trials = Trials()

max_evals = 10000  # some big number of total evaluations
step = 20  # how many to run before saving
for i in range(1, max_evals + 1, step):
    # fmin runs until the trials object has max_evals elements in it, so it can do evaluations in chunks like this
    best = fmin(objective, space, algo=tpe.suggest, trials=trials, max_evals=i)
    # each step 'best' will be the best trial so far
    print(best)
    # each step 'trials' will be updated to contain every result
    # you can save it to reload later in case of a crash, or you decide to kill the script
    pickle.dump(trials, open("my_results.pkl", "wb"))

    # you can also have some logic here to check if `best` hasn't improved in the last X iterations and end early
    # or end after T amount of time has elapsed
```

**Comment by Sanchita-P** [comment]
@bjkomer , Got it! This makes a lot of sense! :) I shall keep this in mind for my future work... (I wished I had asked beforehand...)
Thanks!

**Comment by Sanchita-P** [comment]
@bjkomer - I tried your earlier suggestion, but it didn't improve the problem. 
I did 100 different iterations of the above code - with 1000 evaluations in each (took 3 days to run). The hyperparameters, model performance, and feature importance is very different across all of them! :/ The performance (ROC AUC) swings between 0.7 and 0.8. 
Could you help me figure out what else could be the problem?

Thanks!

**Comment by bjkomer** [comment]
It's possible that your task has a lot of inherent variability based on the hyperparameters and for the size of the space 1000 evaluations is not enough for convergence to a similar model. With 9 hyperparameters, even if there were only 3 possible values for each, the search space would consist of 3^9 = 19,683 possible configurations. Considering the possible values are infinite (continuous), only an extremely small fraction of the space can be evaluated. That being said, hyperopt should still hopefully produce better configurations over time. It's not a gradient based method, so monotonic improvement is not to be expected on each step (though the returned best setting will never get worse).

It sounds like you have a lot of data now so one thing to try would be to plot how performance changes based on each hyperparameter. There may be some insights there to help narrow the search space.

Another check would be to see if there is a trend of improvement over time and whether it looks like convergence will happen anytime soon. The `Trials` object contains every score and parameter setting for each step so you can make a plot of the loss over time like this one shown below (I believe this example had about 65 hyperparameters. Despite being extremely noisy there is still improvement over time, but also no sign of convergence):

![image](https://user-images.githubusercontent.com/1361151/77859509-e3ae9c00-71d7-11ea-873a-af33fbd3a117.png)

**Comment by Sanchita-P** [comment]
Hi @bjkomer , thanks for replying! A few quick clarifications - 
1. _to plot how performance changes based on each hyperparameter_ - here, are you suggesting that I keep all other hyperparameters as default and then check how the performance changes as I vary one hyperparameter? For eg. keep all parameters apart from max_depth as default and vary max_depth to see the change in performance? However, wouldn't the hyperparameter value depend on the combination of other parameters?
2. I tried plotting the suggested graph using `main_plot_vars(trials)` but the plot is very strange - how did you do your plots? I am guessing that you can do both using trials- your above-mentioned graph and graph for performance vs hyperparameter value for each hyperparameter across the different evaluations?

**Comment by bjkomer** [comment]
1. I meant just use the data you already have. For example, for max_depth you can do a scatter plot with max_depth on the x-axis and performance on the y-axis. Each data point will have completely different values for the other parameters, but you may still be able to see some structure in it. I've never used `main_plot_vars(trials)` before, but it might be doing something similar but with the axes flipped. If there are too many evaluations to make sense of a scatter plot, could also try looking at it as a heatmap instead, with frequency of occurrence as the colour.

2. Looking at your loss history graph, it looks like there is improvement up until 500 evals, and then it jumps back to the start and improves again over the next 500 evals. Very strange...

**Comment by Sanchita-P** [comment]
Thanks for replying, @bjkomer :) I know figuring out the solution to this isn't straightforward and I appreciate your patience and support!

1. For 1, I extracted the information for all the variables. I looked at the graphs, and apart from min_child_samples, there is no clear relationship between the model performance and the hyperparameter value. Refer to the graphs below for information. 

2. The lack of improvement is troubling. What else can we try for this?

The orange line represents the best fit linear line and the green represents the best fit polynomial line (degree = 2)
![image](https://user-images.githubusercontent.com/47073198/77864678-f948b380-7221-11ea-8654-e78f67f0304f.png)

**Comment by bjkomer** [comment]
1. The lack of a relationship between individual parameter values and performance could mean that the parameters are highly dependent on combinations in complex ways. Though I still find it strange that they are mostly flat, I would have expected at least some kind of independent relationship (like in min_child_samples) even if the relationship isn't strong. Another reason the results could look like this is if the performance is not deterministic based on the hyperparameters. If you train two models with the exact same hyperparameters, do you always get the same performance? If there is a lot of variance within the same configuration I can seen the results looking like this. It will make it much harder to optimize hyperparameters if this is the case.

2. You said you ran 100 iterations of the 1000 eval code, do they all look like that? Is there a discontinuity at 500 evals in all of them? If so, could be a bug somewhere. Looking at just the first 500 evals that looks to me like improvement (although slow). If the same hyperparameters can produce different performance (due to random seed) this could also make it very hard for performance to improve over time, since each estimate is very noisy, but hyperopt is treating them as exact. If this is the case, one thing that might help is to train the model a few times for each eval of the objective function and take the mean (though this will add a lot more computation).

**Comment by Sanchita-P** [comment]
Hi @bjkomer , apologies for the delay in responding. Find below the response to your comments/ questions - 
1. If we train the model with exactly the same hyperparameters, we always get the **same performance** if we're using the same train and test sets
2. Yes, I did 100 iterations of the 1000 eval code. Here, I was varying the seed used to create the train and test splits. Unfortunately, I didn't store the trial objects for all of them. I did a quick check by running a new iteration with a different seed using the 1000 evaluations. From this, we see the same problem - resetting at 500 evaluations. 
![image](https://user-images.githubusercontent.com/47073198/78041175-4df64680-7368-11ea-8e51-23299169087a.png)
3. Across the 100 iterations (different seed for train and test each time), we saw that the model performance varies a lot and hyperparameters vary. The variation across the iterations: AUC range of 0.7 to 0.8, different hyperparameters and feature importance. In the second point, are you suggesting that in each evaluation, we should perform multiple iterations? That makes it more complicated than it should be, doesn't it? I think I am missing the reasoning and methodology here. Apologies, but could you elaborate? For a given set of hyperparameters, the model always returns the same predictions. 
4. Another concerning thing here is that hyperparameter tuning is not improving the performance of my model (an AUC of 0.760 with default parameters vs 0.766 with hyperparameter tuning)

I wonder what else we can try or what obvious mistake I could be making!

**Comment by bjkomer** [comment]
1. This is very good news! Though I am now curious how much variance there is for the same hyperparameters on different train/test splits. For example, if hyperparameter configuration X does better than Y on train/test split A, does that also mean that X will do better than Y on train/test split B? If not, then a lot of the variation could still be due to noise.

2. From a quick glance, I don't see the same kind of resetting here. Looks to me like there is improvement over the first 200 evals, and then it flattens out or improves much slower from there. Will be easier to see with some kind of trendline. I don't think there is a resetting bug.

3. The suggestion for multiple iterations per evalutation was only for if the model can give different predictions for a given set of hyperparameters, which looks to be not the case (the reasoning was a mean value would be more useful for optimization than a random sample).

4. It seems to me like your model might just be very robust to different hyperparameter settings. I was originally under the impression that just the best result returned was between 0.7 and 0.8, but it looks like literally every combination tried lands in that range, including the first 20 evals which are completely random. I don't know what your application area is or how big of a dataset you have, but how significant is the range of 0.7 to 0.8? (in contrast, datasets/models I've used in the past have varied from ~10% accuracy to ~90% accuracy based on the hyperparameters). One way to test the inherent noise of your system would be to try your trained model with fixed hyperparameters on different splits of the test set and see how much the score varies.

As for other things to try, I only can think of a couple more suggestions. You could cut down the search space by making min_child_samples less than 50, since it seems like values above this tend to do worse regardless of what other parameters are chosen. If you have the time, you could try running a single optimization for more evals (e.g. 10,000), though from what I am seeing, that may not give too much improvement. Can go from where you left of, or even concatenate multiple trials objects you have saved together and start from ever further along (though if they came from different train/test splits, that may not work the greatest).

Hopefully that is helpful, best of luck on your project!

**Comment by Sanchita-P** [comment]
@bjkomer , apologies for the delay in responding. I'd like to first thank you for your help, it's been a wonderful experience working through this with you! :)

Quick update on 2 experiments I tried - 
1. Doing the 100 iterations of different train-test splits without hyperparameter tuning (i.e default LightGBM parameters) - We see a similar range of AUC (0.7 to 0.8).
2. Having the same train-test split and then using different seeds in the hyperopt process - your comments inspired me to check this. In 20 different iterations (each with 1000 evaluations), we see a range of 0.75-0.77 with the distribution as 0.75: 3 times, 0.76: 14 times and 0.77:3 times. There is a huge range in hyperparameters as shown below - 
![image](https://user-images.githubusercontent.com/47073198/79157494-9718b280-7dcc-11ea-98db-93a740e9b94e.png)

Based on this, it seems like the choice of hyperparameters doesn't really make any difference to the model! I am not sure why this is happening. One thing I thought of was the methodology I followed for hyperparameter tuning - 
1. I am not exactly tuning num_iterations using Hyperopt. Instead, in the hyperparameter tuning process, I use early stopping to determine the number of iterations (Value:X). I return this and then use 1.1*(Value:X) in the final training process. I did this to prevent overfitting... Do you think this could cause any problems?
2. While seeing the best hyperparameters, I see that there are combinations of hyperparameters that don't make sense, 'max_depth': 4 and 'num_leaves': 83. Based on traditional understanding, if the maximum depth is 3, the maximum number of leaves possible would be 8... I am obviously missing something here!

After your responses, I think we can bring this discussion to an end. Again, thanks a lot! :)

**Comment by bjkomer** [comment]
I agree that it seems like the choice of hyperparameters doesn't make much difference to the model. While this means `hyperopt` unfortunately won't help you much here, it does indicate that the model you are using is a very robust one. An ideal model is one where you don't need to worry about its hyperparameters. That being said, it is possible that a different model could achieve better scores, and you are just near the limit of what this one can do.

1. This sounds like a good idea to me, I don't think it would cause any problems.

2. I'm not too familiar with LightGBM, but I did some looking around it seems like you are only supposed to set one of `max_depth` or `num_leaves`. If `num_leaves` is set, it actually ignores the `max_depth` parameter! I got this info from skimming [this](https://towardsdatascience.com/lightgbm-vs-xgboost-which-algorithm-win-the-race-1ff7dd4917d) blog post. This definitely explains why that parameter in particular is having no effect. If anything it would just confuse hyperopt if it sees spurious correlations with that parameter to performance. That blog post also mentioned at the end how they saw little improvement when tuning hyperparameters, adding more evidence to the theory that LightGBM is just very robust to its parameters or has good defaults.

It's definitely been a great experience working through this problem. I'm learning lots as well :)

**Comment by Hui-Li5126** [comment]
I stumbled across this discussion and tried this searching code on my dataset with lightGBM with 300 evals, it didn't outperform my default XGBoost for my problem to classify who to send email. I glanced the output hyperparameters, a big chunk of them weren't as deep as XGBoost. I also start to think those hyperparameters might not be the best to experiment on.

**Comment by Severac** [comment]
Thank you very much for your insights bjkomer
In your code, instead of max_evals=i, shouldn't it be max_evals = step ?

>     best = fmin(objective, space, algo=tpe.suggest, trials=trials, max_evals=i)

**Comment by bjkomer** [comment]
@Severac good catch, it should actually be `max_evals=i+step` to include the last set.

**Comment by Severac** [comment]
> @Severac good catch, it should actually be `max_evals=i+step` to include the last set.

Thanks,
I have the same issue of multiple identical attempts, with my former code that is the equivalent of yours with a step of 1 (which allowed me to save results at each step : more flexible to interrupt the work).
Now I've retried with a step of 30, I'll see if there are still identical attempts and better results

**Comment by Severac** [comment]
Well, I still have repetitions : on 30 runs, 7 have exact same hyper parameters (which are the best hyper parameters from previous runs saved to a trial pickle file from which I started)

**Comment by aartdvark** [comment]
I'm just here to say that I have run into the exact same problem, only that I ran the same hyperparameters with RandomizedSearchCV from the sklearn library and sklearn is giving me much better results. Hyperopt stops improving at around 0.42 (F1 score) and random search easily finds hyperparameters that score 0.63. Graphs below show slightly different hyperparamaters as this is my last run, but I ran the two tuning algorithms with identical params before and same result.

These are results for hyperopt:
![image](https://user-images.githubusercontent.com/42552319/119702347-44be1800-be4d-11eb-8169-24c2c1565df3.png)
![image](https://user-images.githubusercontent.com/42552319/119708722-5656ee00-be54-11eb-9d49-649eb1bff8e0.png)
These are results for RandomizedSearchCV:
![image](https://user-images.githubusercontent.com/42552319/119709317-f90f6c80-be54-11eb-9f48-e960dbdd7a92.png)
